### DISCUSSION TITLE
Error when serializing foreign key model images

### DISCUSSION THREAD

**Discussion by Enzodtz** [encode/django-rest-framework#8363] [Potential Issue]
Hey,

I have a model `Pet` and another model `PetImage`, that has a foreign key to `Pet`. When I'm returning a `Pet` from the model serializer `create` method, I think that it might be trying to serialize all model's `PetImage`s, but it is having a problem with their manager, since it raises the exception: `'RelatedManager' object is not iterable`.

I'm not sure if the cause of the problem is right, but please take a look on the serializer code, so I can explain it better:

```python
class PetSerializer(serializers.ModelSerializer):
    images = serializers.ListField(child=serializers.ImageField()) 
    # if I comment the above line, and all the images field references below, the error is not raised anymore 

    class Meta:
        model = Pet
        fields = "__all__"
        extra_kwargs = {
            "user": {
                "default": serializers.CurrentUserDefault(),
            },
            "images": {
                "write_only": True,
            },
        }

    def create(self, validated_data):
        images = validated_data.pop("images")
        pet = Pet.objects.create(**validated_data)

        for image in images:
            PetImage.objects.create(image=image, pet=pet)

        # Also, please notice that the error is only raised in the return line, the above lines run perfectly, and the objects are created
        # That's why I think that the error is when serializing.
        return pet
```

I think that the error is pretty well explained and tracked, but I'm maybe not deep enought into `DRF` to figure out what's happening after that return statement.

Thanks a lot!

###### Additional resources:

<details>
  <summary>Models Code</summary>
  
    class Pet(models.Model):
        name = models.CharField(max_length=50)
        location = models.CharField(max_length=50)
        user = models.ForeignKey(User, on_delete=models.CASCADE)

    class PetImage(models.Model):
        pet = models.ForeignKey(Pet, related_name="images", on_delete=models.CASCADE)
        image = models.ImageField(upload_to="pets")

</details>

<details>
  <summary>View Code</summary>
  
    class PetView(viewsets.ModelViewSet):
        serializer_class = PetSerializer
        queryset = Pet.objects.all()

</details>

**Comment by Enzodtz** [comment]
I manage to track the problem a bit more into the source code: seems that the serialization of the field is maybe not implemented (or I'm using it wrong haha). But please take a look on what I got:

### rest_framework/fields.py: 1670
```python
def to_representation(self, data):
    """
    List of object instances -> List of dicts of primitive datatypes.
    """
    return [self.child.to_representation(item) if item is not None else None for item in data]
```

As you can see, the function is iterating thought `data`, but it shouldn't, since for this manager object you have to use the .all() method before iterating on it. If we change these lines to:

```python
    def to_representation(self, data):
        """
        List of object instances -> List of dicts of primitive datatypes.
        """
        related_manager_type = "<class 'django.db.models.fields.related_descriptors.create_reverse_many_to_one_manager.<locals>.RelatedManager'>"
        if str(type(data)) == related_manager_type: # sorry abt the bad type comparison method
            data = data.all()
            print(data)    

        return [self.child.to_representation(item) if item is not None else None for item in data]
```
The error disappears, and `data` is printed as a list of `PetImage` objects. But the images field is not being serialized as it should, since the request output is:

```json
{
	"id": "<pet_id>",
	"images": [
		null
	],
	"name": "<name>",
	"location": "<location>",
	"user": "<user_id>"
}
```

Maybe this is related to the `to_representation` method also not have been implemented for the image fields?

Also: I set `images` field as `write_only`, but it's still returning it inside the json...

**Comment by Enzodtz** [comment]
Ended up doing this way:

```python
class PetSerializer(serializers.ModelSerializer):
    images = serializers.SerializerMethodField()

    class Meta:
        model = Pet
        fields = "__all__"
        extra_kwargs = {
            "user": {
                "default": serializers.CurrentUserDefault(),
            },
        }

    def get_images(self, pet):
        images = PetImage.objects.filter(pet=pet)

        urls = []
        for image_obj in images:
            urls.append(image_obj.image.url)

        return urls

    def validate(self, data):
        request = self.context.get("request")

        images = request.FILES.getlist("images")

        if len(images) > 0:
            data["images"] = images
            return data
        else:
            raise serializers.ValidationError({"images": "You must submit an image."})

    def create(self, validated_data):
        images = validated_data.pop("images")

        pet = Pet.objects.create(**validated_data)

        for image in images:
            PetImage.objects.create(image=image, pet=pet)

        return pet
```

The problem was that I needed to create `images` field totally from scratch, and that's not ideal. It would be nice if `List + Img` works, and also, if you could use `SerializerMethodField` with an write field.

**Reply by Smixi** [reply]
Well, it does works properly. What you want is to use the "many" arguments  and use your own serializer as you did already:

If i'm correct, listfield is more complex than that, and shouldn't be use like that. It's more likely you will use "many" because of how Listfield are created by DRF Fields. 

```python
class PetSerializer(serializers.ModelSerializer):
    images = serializers.ListField(child=serializers.ImageField()) 
    # if I comment the above line, and all the images field references below, the error is not raised anymore 

    class Meta:
        model = Pet
        fields = "__all__"
        extra_kwargs = {
            "user": {
                "default": serializers.CurrentUserDefault(),
            },
            "images": {
                "write_only": True,
            },
        }
```

If you wanted to have nested serializer with create support, either implement it yourself like you did, or use extra like https://github.com/beda-software/drf-writable-nested

You have an example of how to perform that here : https://www.django-rest-framework.org/api-guide/relations/#nested-relationships
